mayastor/src/bdev/dev/nvmx/controller.rs

```diff
@@ -0,0 +1,1335 @@
+        // side-effects (like active qpairs referring to not existing
+        // controller).
+        assert!(
+            matches!(curr_state, New | Unconfigured),
+            "{} dropping active controller in {:?} state",
+            self.name,
+            curr_state
+        );
+
+        // Inner state might not be yes available.
+        if self.inner.is_some() {
+            let inner = self.inner.take().expect("NVMe inner already gone");
+
+            debug!(
+                ?self.name,
+                "detaching NVMe controller"
+            );
+            let rc = unsafe { spdk_nvme_detach(inner.ctrlr.as_ptr()) };
+
+            debug!(
+                ?self.name,
+                "stopping admin queue poller"
+            );
+            inner.adminq_poller.stop();
+
+            assert_eq!(rc, 0, "Failed to detach NVMe controller");
+            debug!(
+                ?self.name,
+                "NVMe controller successfully detached"
+            );
+        }
+
+        unsafe {
+            Box::from_raw(self.timeout_config.as_ptr());
+        }
+    }
+}
+
+extern "C" fn aer_cb(ctx: *mut c_void, cpl: *const spdk_nvme_cpl) {
+    let mut event = spdk_nvme_async_event_completion::default();
+
+    if !nvme_cpl_succeeded(cpl) {
+        warn!("AER request execute failed");
+        return;
+    }
+
+    event.raw = unsafe { (*cpl).cdw0 };
+
+    let (event_type, event_info) = unsafe {
+        (event.bits.async_event_type(), event.bits.async_event_info())
+    };
+
+    info!(
+        "Received AER event: event_type={:?}, event_info={:?}",
+        event_type, event_info
+    );
+
+    // Populate namespaces in response to AER.
+    if event_type == NvmeAerType::Notice as u32
+        && event_info == NvmeAerInfoNotice::AttrChanged as u32
+    {
+        let cid = ctx as u64;
+
+        match NVME_CONTROLLERS.lookup_by_name(cid.to_string()) {
+            Some(c) => {
+                let mut ctrlr = c.lock().expect("lock poisoned");
+                info!(
+                    "{}: populating namespaces in response to AER",
+                    ctrlr.get_name()
+                );
+                ctrlr.populate_namespaces();
+            }
+            None => {
+                warn!(
+                    "No NVMe controller exists with ID 0x{:x}, no namespaces rescanned",
+                    cid,
+                );
+            }
+        }
+    }
+}
+
+/// return number of completions processed (maybe 0) or negated on error. -ENXIO
+/// in the special case that the qpair is failed at the transport layer.
+pub extern "C" fn nvme_poll_adminq(ctx: *mut c_void) -> i32 {
+    let mut context = NonNull::<TimeoutConfig>::new(ctx.cast())
+        .expect("ctx pointer may never be null");
+    let context = unsafe { context.as_mut() };
+
+    let rc =
+        unsafe { spdk_nvme_ctrlr_process_admin_completions(context.ctrlr) };
+    if rc < 0 {
+        context.reset_controller();
+    }
+
+    if rc == 0 {
```

Comment:
So we return 1 completion in case of error or >0 completions?

Reply:
The idea is to reset the controller if error occurs upon processing admin completions, and mark poller as busy (return non-zero value) in case of an error, or idle (return zero value) if completions are ok. We re-implement here existing native nvme driver's logic from spdk.